Drop integration tests, add e2e tests with kind

mudler opened the pull request:
Fixes: https://github.com/rancher-sandbox/rancheros-operator/issues/2
Fixes: https://github.com/rancher-sandbox/rancheros-operator/issues/6

Need also to cleanup a bit

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler committed:
Drop integration tests, add e2e tests with kind

Fixes: https://github.com/rancher-sandbox/rancheros-operator/issues/2
Fixes: https://github.com/rancher-sandbox/rancheros-operator/issues/6

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

mudler commented on Makefile:
```diff
@@ -32,21 +32,15 @@ chart:
+e2e-tests:
+	kind delete cluster --name "ros-e2e"
+	kind create cluster --name "ros-e2e"
+	kubectl cluster-info --context kind-ros-e2e
```
gah external_ip succeds locally but it doesn't on the runner, have to check why

Itxaka commented on Makefile:
```diff
@@ -32,21 +32,15 @@ chart:
+	kind delete cluster --name "ros-e2e"
+	kind create cluster --name "ros-e2e"
+	kubectl cluster-info --context kind-ros-e2e
+	cd $(ROOT_DIR)/tests && EXTERNAL_IP=$(shell kubectl get nodes -o json | jq -r '.items[].status.addresses[] | select(.type == "InternalIP").address') ginkgo -r -v ./e2e
```
no need for external tools:
`kubectl get nodes -o jsonpath='{.items[].status.addresses[?(@.type == "InternalIP")].address}'`

Itxaka commented on tests/e2e/e2e_suite_test.go:
```diff
@@ -0,0 +1,85 @@
+
+	chart = os.Getenv("ROS_CHART")
+	if chart == "" {
+		Fail("No chart provided, skipping tests")
```
Skip instead of Fail

mudler replied:
Well, tests depends on it!

Itxaka replied:
oh well, then we should change the text :D

Itxaka commented on tests/e2e/e2e_suite_test.go:
```diff
@@ -0,0 +1,85 @@
+		err := sut.RunHelmBinaryWithCustomErr("-n", "cert-manager", "install", "--set", "installCRDs=true", "--create-namespace", "cert-manager", "https://charts.jetstack.io/charts/cert-manager-v1.5.3.tgz")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = kubectl.WaitForPod("cert-manager", "app.kubernetes.io/instance=cert-manager", "cert-manager-cainjector")
```
:heart_eyes:

Itxaka replied:
oh wait, this is ours LOL

I thought we were using an upstream lib to control that :D

Itxaka commented on tests/sut/kubectl.go:
```diff
@@ -0,0 +1,683 @@
+
+// checkPodTerminateLabelFilter checks is the pod status is terminated
+func (k *Kubectl) checkPodDeleted(namespace string, name string) (bool, error) {
+	out, err := runBinary(kubeCtlCmd, "--namespace", namespace, "get", "pod", name)
```
isnt there a golang way of getting this info with a client instead of forking to binaries?

mudler replied:
yes but wouldn't be anymore an e2e test..plus we would end up pinning deps that we don't really need. Here we can later have a matrix for kubernetes versions indipendent from the code

Itxaka commented on tests/sut/kubectl.go:
```diff
@@ -0,0 +1,683 @@
+	})
+}
+
+// SecretExists returns true if the pod by that name is in state running
```
```suggestion
// SecretExists returns true if the secret by that name is in state running
```

Itxaka reviewed (commented):
wow, I mean, its cool but a bit overboard with a lot of functions that we dont need. Looks good to me thought, just a couple of nits and a (future?) implementation issue.

mudler commented:
> wow, I mean, its cool but a bit overboard with a lot of functions that we dont need. Looks good to me thought, just a couple of nits and a (future?) implementation issue.

yup, I'm indeed not sure if just cut what we need here or just port it as a standalone lib. At the end, I see us reusing that code parts in the e2e tests and we could just reuse that code to quickly ramp-up without have to re-implement such helpers ourselves.

mudler committed:
Move helpers to standalone lib

- Also add nginx as ingress as is more compliant to kind
- Move out some common code pieces
- Address points from review

Signed-off-by: Ettore Di Giacinto <edigiacinto@suse.com>

ldevulder commented on tests/e2e/machineregistration_test.go:
```diff
@@ -0,0 +1,75 @@
+
+var _ = Describe("MachineRegistration e2e tests", func() {
+	k := kubectl.New()
+	Context("registration", func() {
```
Is `Context` really useful here? As there is only this one (or maybe more are planned?).

mudler replied:
I hoped there was more in the pipeline for that change at hand :sweat_smile: but this will get bigger I'm pretty sure..

ldevulder commented on tests/e2e/e2e_suite_test.go:
```diff
@@ -0,0 +1,140 @@
+	RunSpecs(t, "ros-operator e2e test Suite")
+}
+
+func waitNamespace(namespace, label string) {
```
Maybe this function could be added in `https://github.com/rancher-sandbox/ele-testhelpers/blob/main/kubectl/kubectl.go`?

ldevulder replied:
Ok, I can do it.

ldevulder reviewed (commented):
2 comments a little bit late :) but as my bootstrapping tests conversion code to ginkgo as some similarities I will try to use the https://github.com/rancher-sandbox/ele-testhelpers lib.

In my current version I use some code from Epinio, but this lib could be more powerful. My main concern is to use `kubectl` command and not API, which looks like to be the case in the lib.

mudler commented:
> 2 comments a little bit late :) but as my bootstrapping tests conversion code to ginkgo as some similarities I will try to use the https://github.com/rancher-sandbox/ele-testhelpers lib.
> 
> In my current version I use some code from Epinio, but this lib could be more powerful. My main concern is to use `kubectl` command and not API, which looks like to be the case in the lib.

Yes, that is done in purpose to avoid to pinpoint kubernetes api versions in all the code consuming it downstream. It uses also standalone kubectl so we can just setup matrix tests without caring about the api versions attached to the test code

ldevulder commented:
> It uses also standalone kubectl so we can just setup matrix tests without caring about the api versions attached to the test code

Yes, and it's better to use the command like a real user instead of the API for e2e tests IMO :-)
